lib/api/bucketPutVersioning.js

```diff
@@ -0,0 +1,84 @@
+import async from 'async';
+import { parseString } from 'xml2js';
+
+import metadata from '../metadata/wrapper';
+import services from '../services';
+
+/**
+ * Format of xml request:
+
+ <VersioningConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
+    <Status>VersioningState</Status>
+    <MfaDelete>MfaDeleteState</MfaDelete>
+ </VersioningConfiguration>
+
+ Note that there is the header in the request if setting MfaDelete:
+    x-amz-mfa: [SerialNumber] [TokenCode]
+ */
+
+/**
+ * Bucket Put Versioning - Create or update bucket Versioning
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+export default function bucketPutVersioning(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'bucketPutVersioning' });
+
+    const bucketName = request.bucketName;
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'bucketPutVersioning',
+        log,
+    };
+
+    return async.waterfall([
```

Comment:
Could we have each function defined out of the waterfall to avoid allocating it every time and not doing any v8 optimisation?